This feature is to create an eslintcache that can be shared across de… (#114)
eslint/rfcs: designs/2023-relative-cache/README.md

- eslint-helpers.js: add shareableCache variable (set to false) to the processOptions
- lib/options.js: Add an option 'shareable-cache' of type Boolean with a nice description for people to read: By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache.

### Changing cache file serialization
- lib/eslint/flat-eslint.js: Define the shearableCache property at the top, with the other properties, as a boolean

> - lib/eslint/flat-eslint.js: Around line 873, where the flat-eslint.js file is saving the cache in the cachefile, if processOptions.shareableCache is set to true, pass a PARTIAL path to the lintResultCache.setCachedLintResults() function instead of the full path. 

Author: Thank you!  You're right. After some testing, I found some things I missed in my first design go-around. I will be revising accordingly.

> - lib/eslint/flat-eslint.js: Around line 873, where the flat-eslint.js file is saving the cache in the cachefile, if processOptions.shareableCache is set to true, pass a PARTIAL path to the lintResultCache.setCachedLintResults() function instead of the full path. 

Author: Please recheck the section "Changing cache file serialization" as I think I have addressed your concerns. Thank you!

```

### Suggested updated cache file

Here is a suggested updated cache file, with relative paths instead of full paths. The `.` is the current working  
directory. Typically, eslint is always run from the same place, so using a relative path should not cause a problem.

Note that the obscure hash value is still there, and is now Ak0sovs. This hash value will need to be calculated using
relative paths only. 

``` 
[
  {
    "./src/formatter.ts": "1",
    "./src/vite-env.d.ts": "2"
  },
  {
    "size": 2584,
    "mtime": 1695823503788,
    "results": "3",
    "hashOfConfig": "4"
  },
  {
    "size": 75,
    "mtime": 1695814197580,
    "results": "5",
    "hashOfConfig": "4"
  },
  {
    "filePath": "6",
    "messages": "7",
    "suppressedMessages": "8",
    "errorCount": 0,
    "fatalErrorCount": 0,
    "warningCount": 0,
    "fixableErrorCount": 0,
    "fixableWarningCount": 0
  },
  "Ak0sovs",
  {
    "filePath": "9",
    "messages": "10",
    "suppressedMessages": "11",
    "errorCount": 0,
    "fatalErrorCount": 0,
    "warningCount": 0,
    "fixableErrorCount": 0,
    "fixableWarningCount": 0
  },
  "./src/formatter.ts",
  [],
  [],
  "./src/vite-env.d.ts",
  [],
  []
]
```

> ### Adding the command line parameter

Reviewer: I'm assuming this will also be a new [`ESLint` constructor](https://eslint.org/docs/latest/integrate/nodejs-api#-new-eslintoptions) option, so we should mention that as well.

- `lib/cli.js`: Add the `shareableCache` option, defaulted to false, in the `translateOptions `function. 

### Changing cache file serialization
- `lib/cli-engine/lint-result-cache.js`: Add the properties `cwd` and `shareableCache` to the `LintResultCache` class. `cwd` is a string, the current working directory. `shareableCache` is a boolean, the result of the user's passed in command line parameter. Use these values in two main places: 
    1. Whenever we store or retrieve file results in the cache, use the results of the new class function, `getFileKey(filePath)`. If `shareableCache` is false, it uses a full file path as the key. If `shareableCache` is true, uses a relative file path. 
    2. Update the function `hashOfConfigFor`, so that when the configuration object is hashed into a hash key and put into the eslintcache, if `shareableCache` is set to true, the file paths hashed are all relative paths. Implementation detail:
```js
function hashOfConfigFor(config, cwd) {
    if (!configHashCache.has(config)) {

>         // replace the full directory path in the config string to make the hash location-neutral

Reviewer: Since this feature will only be implemented for the new config system, I believe there is no need for this because there are no absolute paths in configs.

Reviewer: A place where we do have absolute paths is [`LintResult#filePath`](https://eslint.org/docs/latest/integrate/nodejs-api#-lintresult-type), so that's something we'll need to convert from absolute to relative and vice versa when cache entries are stored and retrieved.

Reviewer: > A place where we do have absolute paths is [`LintResult#filePath`](https://eslint.org/docs/latest/integrate/nodejs-api#-lintresult-type), so that's something we'll need to convert from absolute to relative and vice versa when cache entries are stored and retrieved.

Or, we could omit `LintResult#filePath` property from cache entries and add it back when cache entries are retrieved.

Reviewer: A reminder that the question about how [`LintResult#filePath`](https://eslint.org/docs/latest/integrate/nodejs-api#-lintresult-type) will be handled should be addressed in this RFC.

        // implementation TBD; please note configHashCache is a complex object with over 100 fields.        
    }
``` 
- lib/cli-engine/cli-engine.js:  Update the constructor call to LintResultCache to add the new parameters cwd and shareableCache. 

## Documentation

Add the new CLI flag to the CLI documentation. The new cli flag will be called shareable-cache.

## Drawbacks

Adding a new CLI flag adds some complexity for the end user. They will need to understand why the feature exists. 

Furthermore, defaulting to the existing behavior does help make upgrading less painful, but it is possible most people expect a shareable cache by default. This may cause confusion.

## Backwards Compatibility Analysis

The new flag, shareable-cache, passed in on the command line, will default to false. Unless people explicitly set it to true, then the old algorithms will
be used. Users of the eslint cache will still need to regenerate their cache no matter how they set this flag, because the current implementation forces a cache rebuild if there are

> any configuration changes. Adding a new property counts as a change. 

Reviewer 2: Can you clarify the last sentence? Is _a new property_ some field in the cache, in the config, or what exactly?

Reviewer: Any new version of ESLint invalidates cache entries stored by a previous version, so I think this is not a concern.

https://github.com/eslint/eslint/blob/d2d83e045ad03f024d1679275708054d789ebe20/lib/cli-engine/lint-result-cache.js#L50

- `eslint-helpers.js`: add `shareableCache` variable (set to false) to the `processOptions`, and make sure it must be a boolean. 
- `lib/options.js`: Add an option `shareable-cache` of type Boolean with a nice description for people to read: By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache.
- `lib/cli.js`: Add the `shareableCache` option, defaulted to false, in the `translateOptions `function. 

### Changing cache file serialization

> - `lib/cli-engine/lint-result-cache.js`: Add the properties `cwd` and `shareableCache` to the `LintResultCache` class. `cwd` is a string, the current working directory. `shareableCache` is a boolean, the result of the user's passed in command line parameter. Use these values in two main places: 

Reviewer: I think we should clarify what we mean by the current working directory that will be passed to the `LintResultCache` class. Is it `process.cwd()`, or `cwd` that was passed to ESLint/FlatESLint constructor? In the latter case, is there a way to pass `cwd` to `file-entry-cache` or we could end up calculating paths differently?

Reviewer: Another question is does `file-entry-cache` support relative paths? All tests seem to be with absolute paths:

https://github.com/jaredwray/file-entry-cache/blob/master/test/specs/cache.js

Author: It will. I am just going to push up the code review sometime today. I have a complete and working branch on this task, and most of  your questions wuold be resolved by the code review/better answered looking at the code changes.

Reviewer 3: Don't forget to answer the question regarding what `cwd` is from the previous comment.

Author: > Is it `process.cwd()`, or `cwd` that was passed to ESLint/FlatESLint constructor?

When created in cli-engine.js, it's options.cwd. It is also referred to as 'cwd' in the code. 
When created in flat-eslint.js, it's processedOptions.cwd.

Author: Ok. I have updated the code, and the flat-eslint.js file no longer exists in my code branch. I see a lot of the code duplication around parameter parsing has been removed. I am updating my RFC accordingly.

Reviewer: I think the question whether `file-entry-cache` supports relative paths and what happens when `cwd` passed to ESLint is different than `process.cwd()` hasn't been addressed.

Reviewer: To clarify the problem with cwd: When we store lint results for a file, `file-entry-cache` reads that file to get the data (metadata or content, depending on the cache strategy) it will use next time we use cache to determine if the file has been modified in the meantime and thus if its cache entry should be invalidated. So, when we pass a file path, `file-entry-cache` needs to find that file on the disk. If the path is relative, the question is how `file-entry-cache` resolves it. If we pass a path that is relative to `options.cwd`, and `file-entry-cache` resolves it relative to `process.cwd()` (likely), then it won't find the file when `options.cwd` is different than `process.cwd()`.

Another important question is whether `file-entry-cache` officially supports relative paths, or maybe that works just by chance but could break without notice.

Reviewer 3: @jaredwray 👋 Would you be able to help us answer this question: does `file-entry-cache` expect or require absolute paths? Is there any reason you could think of that would prevent it from using relative paths?

Reviewer 4: @nzakas - I dont believe it would matter if you use relative or absolute. Want me to create some tests to validate?

Reviewer 3: @jaredwray that would be awesome, thank you. 🙏

Reviewer 4: great. I will add in some unit tests over the next couple days to the main project (https://github.com/jaredwray/file-entry-cache) to make sure it doesnt matter. 

From what I am gathering we want to be able to do:

```js
const path = require('path');
const fileEntryCache = require('file-entry-cache');
const relativePath = './relative-cache';
const absolutePath = path.join(process.cwd(), relativePath);
const cacheRelative = fileEntryCache.create('cacheName', [relativePath]);
const cacheAbsolute = fileEntryCache.create('cacheName', [absolutePath]);

//then we run tests to validate each!

```

Is this what we want to verify or something more?

Reviewer 3: We'd like to verify that we can both write to and read from the cache with both absolute and relative paths.

The fear from the thread above, is that we may be relying on unintentional behavior that could change unexpectedly. So by adding these tests, we can verify both that this works and that it won't accidentally stop working in the future.

Reviewer 3: @mdjermanovic do you have an example that would satisfy your question?

Reviewer 4: @nzakas  - I am writing the tests that will check that you can read and write with relative and absolute to validate this over the next 72 hours and will post the results here.

Reviewer 3: Awesome, thank you!

Reviewer 3: @jaredwray just checking back to see if you had a chance to look into those tests?

Reviewer 4: I wrote some tests using `cwd()` and it seems to be working just great. Are you seeing any issues? If so, can you give me an easy example of how to reproduce so we can fix it?

Reviewer 4: Here is a run down from our tests:
* create (works with both)
* createFromFile (works with both)

The issue is that the keys are absolute paths so what happens is that when you do something like:
* .getUpdatedFiles(files);
* .hasFileChanged(files[0]);

etc... These are all based on the absolute path for the key. Does that make sense?

Reviewer 4: @nzakas Ok. So in this case you would want to pass in something like `../folder/file_name.txt` instead of the full path. Correct?

Reviewer 4: Yep. It requires an absolute path. I could work to do a release where it doesn't. Is that the best option for you all?

```javascript
  it('getFileDescriptor with relative paths', function () {
    cache = fileEntryCache.create('testCache');
    var file = '../fixtures/' + fixtureFiles[0].name;
    cache = fileEntryCache.createFromFile('../fixtures/.eslintcache');
    expect(cache.getFileDescriptor(file).notFound).to.be.false;
    cache.reconcile();
    deleteFileSync(file);
    expect(cache.getFileDescriptor(file).notFound).to.be.true;
  });
```

This failed on the first expect as the file should exist.

Reviewer 4: @nzakas - I pushed a new pull request that detects if it is absolute or relative and then does the conversion on `removeEntry` and `getFileDescriptor` which are the two main culprits. Can you all test if this fixes it? If it does, then I am happy to do a release. 

https://github.com/jaredwray/file-entry-cache/pull/46

Reviewer 3: For our purposes, what we're looking for is the ability to reference files relative to the cache file itself rather than from the cwd.

For example, if we have a `.eslintcache` file, we'd like to be able to say `getFileDescriptor("foo/bar.js")` and have that understood to be relative to the `.eslintcache` location. 

The goal is to have the cache file continue to work even if we move it to a different system where the directory structure above it might be different but the directory structure below it is the same.

Reviewer 4: makes sense. I am wondering if we add in a rootPath parameter that you can set or change if you move the file and it keeps the consistency? Something along the lines of:

```javascript
cache = fileEntryCache.create('testCache');
var cache = fileEntryCache.createFromFile('/fixtures/.eslintcache');
var info1 = cache.getFileDescriptor('foo/bar.js');
cache.reconsile();
cache.rootPath = '/fixtures/.eslintcache1';
cache.reconsile();  // this updates the base path so that you can 
var info2 = cache.getFileDescriptor('foo/bar.js');
expect(info1).to.equal(info2);
```

Reviewer 3: I'm curious if we need a root path at all? As in, do you need to store anything other than the relative path as it was passed?

Reviewer 4: I was assuming that the path would change to a new one so we would need to know where the new one would be.

Reviewer 4: do you want to setup some time to do a video call? I can do it this weekend if needed as I don't want to be the bottleneck on this solution. :-)

Reviewer: Here's the list of methods we're using:

* API `create()`
* `FileEntryCache#getFileDescriptor()`
* `FileEntryCache#reconcile()`

Reviewer: @jaredwray We'd like to make cache files shareable between different systems, where the project is possibly cloned/copied into a different directory.

For example, currently, if someone has the following directory structure and makes a cache file:

```
/
└── projects/
    └── my-app/
        ├── .eslintcache
        ├── eslint.config.js
        └── my-file.js
```

the cache entry for `my-file.js` is saved with `"/projects/my-app/my-file.js"` key.

Then, if `my-app/` is copied to another system into directory `work/`:

```
/
└── work/
    └── my-app/
        ├── .eslintcache
        ├── eslint.config.js
        └── my-file.js
```

the cache file is unusable because there's no `/projects/my-app/my-file.js` file on that system.

We've had multiple requests to address this by using relative paths instead of absolute paths as keys in the cache file:

* https://github.com/eslint/eslint/issues/16493
* https://github.com/eslint/eslint/issues/16741
* https://github.com/eslint/eslint/issues/14571
* https://github.com/eslint/eslint/issues/17514

Reviewer: A detail I think we're still unsure about is which location the paths should be relative to - what would make the most sense for ESLint users, and then what could `file-entry-cache` support?

Possible candidates:

1. `process.cwd()`
2. `cwd` option passed to ESLint (the same as `process.cwd()` when ESLint is used via CLI)
3. Location of the cache file
4. Location of the `eslint.config.js` file (`basePath`)

Option 1 (and consequently option 2 when ESLint is used via CLI) could be problematic because ESLint can be run from different locations so file paths in the cache might end up being inconsistent.

@nzakas per https://github.com/eslint/rfcs/pull/114#discussion_r1514962901, seems that you think that option 3 makes the most sense?

Reviewer: > ```js
> cache = fileEntryCache.create('testCache');
> var cache = fileEntryCache.createFromFile('/fixtures/.eslintcache');
> var info1 = cache.getFileDescriptor('foo/bar.js');
> cache.reconsile();
> cache.rootPath = '/fixtures/.eslintcache1';
> cache.reconsile();  // this updates the base path so that you can 
> var info2 = cache.getFileDescriptor('foo/bar.js');
> expect(info1).to.equal(info2);
> ```

We don't intend to support moving the cache file during the execution. What we'd like to support is something like this:

Step 1: Run eslint in directory `/foo` and store data for file `bar.js` (`/foo/bar.js`) in the cache

```js
const fileEntryCache = require("file-entry-cache");

const cache = fileEntryCache.create("/foo/.eslintcache", undefined, /* usecheckSum = */ true);

const fileDescriptor = cache.getFileDescriptor("bar.js");

// store some data
fileDescriptor.meta.results = { something: 42 };

cache.reconcile();
```

Step 2: Copy the entire content (including the cache file) of `/foo` into `/baz`

Step 3: Run eslint in directory `/baz` and use the stored data for file `bar.js` (now `/baz/bar.js`) from the cache.

```js
const fileEntryCache = require("file-entry-cache");

const cache = fileEntryCache.create("/baz/.eslintcache", undefined, /* usecheckSum = */ true);

const fileDescriptor = cache.getFileDescriptor("bar.js");

// use the data
if (!fileDescriptor.changed) { // fileDescriptor.changed should be `false` because the content of `bar.js` hasn't changed
    console.log(fileDescriptor.meta.results); // should be { something: 42 }
}
```

Reviewer 4: This makes sense and I plan to have a pull request on Monday (4/1) that can handle this scenario so you can validate with it first. Is that ok?

Reviewer 2: > Ok. I have updated the code, and the flat-eslint.js file no longer exists in my code branch. I see a lot of the code duplication around parameter parsing has been removed. I am updating my RFC accordingly.

In ESLint 9, eslint.js has been renamed to legacy-eslint.js and flat-eslint.js is now named eslint.js.

Reviewer 2: @jaredwray are you still interested in submitting a pull request? Just let us know if you need any help.

Reviewer 4: @fasttime - I made the changes to support the relative path and wanted to confirm that is the change you need to make it work with file-entry-cache.

Reviewer 4: @fasttime - https://github.com/jaredwray/file-entry-cache/pull/46

Reviewer 2: > @fasttime - [jaredwray/file-entry-cache#46](https://github.com/jaredwray/file-entry-cache/pull/46)

Thanks! Now it would be great if someone could draft a PR in ESLint to test these changes.

Reviewer: > jaredwray/file-entry-cache#46

@jaredwray I tried this with passing relative paths from ESLint, and it didn't work for the scenario described in https://github.com/eslint/rfcs/pull/114#discussion_r1532521752. It looks like the change is that, when a relative path is passed to `getFileDescriptor()`, it converts it to an absolute path, but at the end cache file still stores absolute paths?

Reviewer: @nzakas, @fasttime there's still a design question for us: which location the paths should be relative to? (https://github.com/eslint/rfcs/pull/114#discussion_r1532605561)

Reviewer 2: > @nzakas, @fasttime there's still a design question for us: which location the paths should be relative to? ([#114 (comment)](https://github.com/eslint/rfcs/pull/114#discussion_r1532605561))

That should be probably the directory containing the cache file, i.e. option 3 in the comment linked above. Option 4 would also work but with the additional requirement that a config file must exist, which is not needed to work with the ESLint API or to use the cache.

Reviewer 3: I think ideally it would be relative to the location of the cache file, however, if someone is using `--cache-location`, then I think the expectation may be that the paths are relative to cwd?

Reviewer 3: @mdjermanovic @fasttime what do you think of [my last comment](https://github.com/eslint/rfcs/pull/114#discussion_r1616179161)?

Reviewer 2: > I think ideally it would be relative to the location of the cache file, however, if someone is using `--cache-location`, then I think the expectation may be that the paths are relative to cwd?

When `--cache-location` is not specified, the location of the cache file is `cwd` by default. So if there is an expectation that the paths are relative to `cwd` when `--cache-location` is used, the same expectation should be probably there when `--cache-location` is not used. Do you think that `cwd` would be a better base location to store relative paths?

Reviewer 3: > Do you think that `cwd` would be a better base location to store relative paths?

Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.

Reviewer 2: > > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.

> > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.

> > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.

> > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.

I'm asking because when `--cache-location` is not set, the cache file location should be already `cwd`. I think the advantage of keeping the paths relative to the cache file directory is that they don't change if the current directory changes, for example in a monorepo. Do you have a particular scenario in mind where storing paths relative to `cwd` would be advantageous?

Reviewer: Yeah, as @fasttime noticed, when `--cache-location` is not used, the cache file defaults to `.eslintcache` _in cwd_. So if the logic would be to use the location of the cache file when `--cache-location` is not used and cwd when `--cache-location`  is used, then it's basically the same as - always use cwd.

Reviewer 3: I think we should just make a decision and see what happens when people test with it. This PR has been open since October of last year, and it seems like this is the last thing to decide on, so I don't think there's value in drawing this out further when we have a binary decision to make.

So I'd like to propose that we say the paths are always relative to location of the `.eslintcache` file.

Reviewer: > So I'd like to propose that we say the paths are always relative to location of the `.eslintcache` file.

I agree.

    1. Whenever we store or retrieve file results in the cache, use the results of the new class function, `getFileKey(filePath)`. If `shareableCache` is false, it uses a full file path as the key. If `shareableCache` is true, uses a relative file path. 
    2. Update the function `hashOfConfigFor`, so that when the configuration object is hashed into a hash key and put into the eslintcache, if `shareableCache` is set to true, the file paths hashed are all relative paths. Implementation detail:
```js
function hashOfConfigFor(config, cwd) {
    if (!configHashCache.has(config)) {
        // replace the full directory path in the config string to make the hash location-neutral
        // implementation TBD; please note configHashCache is a complex object with over 100 fields.        
    }

> ``` 

Reviewer: Can you clarify this? `config` is eslint configuration for the file, I'm not sure what "replace the full directory path in the config string" refers to.

Reviewer: Oh, I see, in eslintrc mode we're passing here a config array with absolute paths to config files.

If we implement this feature for flat config only (current `FlatESLint` class), will there be a need to change anything in this function?

Author: Yes, because the values of the fields go into the hash, and the values include the full paths.

- lib/cli-engine/cli-engine.js:  Update the constructor call to LintResultCache to add the new parameters cwd and shareableCache. 
- lib/eslint/flat-eslint.js: Update the constructor call to LintResultCache to add the new parameters cwd and shareableCache. 

## Documentation

Add the new CLI flag to the CLI documentation. The new cli flag will be called shareable-cache.

> ## Drawbacks

Reviewer 5: Is there any situation where using a relative path prevents the cache from working depending on where ESLint is run from?

Author: i do not think there would be an issue, but I am not the expert.

Furthermore, defaulting to the existing behavior does help make upgrading less painful, but it is possible most people expect a shareable cache by default. This may cause confusion.

## Backwards Compatibility Analysis

The new flag, shareable-cache, passed in on the command line, will default to false. Unless people explicitly set it to true, then the old algorithms will
be used. Users of the eslint cache will still need to regenerate their cache no matter how they set this flag, because the current implementation forces a cache rebuild if there are
any configuration changes. Adding a new property counts as a change. 

## Alternatives

A CLI tool which translates the existing cache to the current folder structure. This could be any sort of script that will basically do a replace-all on the paths in 
the .eslintcache file and also recalculates the hash so that they are the new location rather than the old. A downside to this is that any tool based on string 
replacement would be fragile at best. 

## Open Questions

Question: Should we make every cache shareable, or keep the shreable-cache flag default to false?

> Answer: We don't know how people are currently using the cache file, so we can't make assumptions about what is and isn't safe to change. Even if we made a shareable cache the default (which would be a breaking change), we'd still need a way to enable the previous behavior in case there is a use case we aren't aware of..

Reviewer 5: If there are no downsides, I think the RFC should propose making this relative cache the default behavior as a future breaking change. A one-time cache file upgrade isn't a big deal and wouldn't be surprising during a major version upgrade. It sounds like the option would still be available if someone needs to opt-out of this behavior, but the RFC hasn't identified any reasons why someone would want to opt-out...

Author: Nzakas gave this answer. I am happy with whatever your team decides. What i have currently working uses a flag, default value false.

1k0siqs. That is the hash-value of a stringified ConfigArray object that contains all the configuration for the entire eslint run. It is there so
that any rule changes or configuration changes will cause a complete eslint recheck on all files. 

```
[
  {
    "/home/USER/git/samplecode/src/formatter.ts": "1",
    "/home/USER/git/samplecode/src/vite-env.d.ts": "2"
  },
  {
    "size": 2584,
    "mtime": 1695823503788,
    "results": "3",
    "hashOfConfig": "4"
  },
  {
    "size": 75,
    "mtime": 1695814197580,
    "results": "5",
    "hashOfConfig": "4"
  },
  {
    "filePath": "6",
    "messages": "7",
    "suppressedMessages": "8",
    "errorCount": 0,
    "fatalErrorCount": 0,
    "warningCount": 0,
    "fixableErrorCount": 0,
    "fixableWarningCount": 0
  },
  "1k0siqs",
  {
    "filePath": "9",
    "messages": "10",
    "suppressedMessages": "11",
    "errorCount": 0,
    "fatalErrorCount": 0,
    "warningCount": 0,
    "fixableErrorCount": 0,
    "fixableWarningCount": 0
  },
  "/home/USER/git/samplecode/src/formatter.ts",
  [],
  [],
  "/home/USER/git/samplecode/src/vite-env.d.ts",
  [],
  []
]

```

### Suggested updated cache file

> Here is a suggested updated cache file, with relative paths instead of full paths. The . is the directory from which 

Reviewer 3: We can use backticks to make this a bit more readable.
```suggestion
Here is a suggested updated cache file, with relative paths instead of full paths. The `.` is the directory from which 
```

> eslint is run. Typically, eslint is always run from the same place, so using a relative path should not cause a problem.

Reviewer 3: "the directory from which eslint is run" - this is typically called the current working directory.

Note that the obscure hash value is still there, and is now Ak0sovs. This hash value will need to be calculated using
relative paths only. 

``` 
[
  {
    "./src/formatter.ts": "1",
    "./src/vite-env.d.ts": "2"
  },
  {
    "size": 2584,
    "mtime": 1695823503788,
    "results": "3",
    "hashOfConfig": "4"
  },
  {
    "size": 75,
    "mtime": 1695814197580,
    "results": "5",
    "hashOfConfig": "4"
  },
  {
    "filePath": "6",
    "messages": "7",
    "suppressedMessages": "8",
    "errorCount": 0,
    "fatalErrorCount": 0,
    "warningCount": 0,
    "fixableErrorCount": 0,
    "fixableWarningCount": 0
  },
  "Ak0sovs",
  {
    "filePath": "9",
    "messages": "10",
    "suppressedMessages": "11",
    "errorCount": 0,
    "fatalErrorCount": 0,
    "warningCount": 0,
    "fixableErrorCount": 0,
    "fixableWarningCount": 0
  },
  "./src/formatter.ts",
  [],
  [],
  "./src/vite-env.d.ts",
  [],
  []
]
```

### Adding the command line parameter

> - conf/default-cli-options.js: add the property 'shareableCache' with a default of false. It should be put in the section with the other cache variables, below cacheStrategy. 
> - docs/src/use/command-line-interface.md: Add an explanation of the 'shareable-cache' variable with the other cache variables: "By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache."
> - eslint-helpers.js: add shareableCache variable (set to false) to the processOptions, and make sure it must be a boolean. 
> - lib/options.js: Add an option 'shareable-cache' of type Boolean with a nice description for people to read: By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache.
> - lib/cli.js: Add the shareableCache option, defaulted to false, in the translateOptions function. 

Reviewer 3: Again, not quite sure "shareable cache" is the right name for this. @mdjermanovic what do you think?

Also, let's use backticks to make things more readable.

```suggestion
- `conf/default-cli-options.js`: add the property `shareableCache` with a default of false. It should be put in the section with the other cache variables, below `cacheStrategy`. 
- `docs/src/use/command-line-interface.md`: Add an explanation of the `shareable-cache` variable with the other cache variables: "By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache."
- `eslint-helpers.js`: add `shareableCache` variable (set to false) to the `processOptions`, and make sure it must be a boolean. 
- `lib/options.js`: Add an option `shareable-cache` of type Boolean with a nice description for people to read: By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache.
- `lib/cli.js`: Add the `shareableCache` option, defaulted to false, in the `translateOptions `function. 
```

Author: Can you suggest a new name? I will use what you like.

### Changing cache file serialization

> - lib/cli-engine/lint-result-cache.js: Add the properties 'cwd' and 'shareableCache' to the LintResultCache class. 'cwd' is a string, the current working directory. shareableCache is a boolean, the result of the user's passed in command line parameter. Use these values in two main places: 
>     1. Whenever we store or retrieve file results in the cache, use the results of the new class function, getFileKey(filePath). If shareableCache is false, it uses a full file path as the key. If shareableCache is true, uses a relative file path. 
>     2. Update the function hashOfConfigFor, so that when the configuration object is hashed into a hash key and put into the eslintcache, if shareableCache is set to true, the file paths hashed are all relative paths. Implementation detail:

Reviewer 3: ```suggestion
- `lib/cli-engine/lint-result-cache.js`: Add the properties `cwd` and `shareableCache` to the `LintResultCache` class. `cwd` is a string, the current working directory. `shareableCache` is a boolean, the result of the user's passed in command line parameter. Use these values in two main places: 
    1. Whenever we store or retrieve file results in the cache, use the results of the new class function, `getFileKey(filePath)`. If `shareableCache` is false, it uses a full file path as the key. If `shareableCache` is true, uses a relative file path. 
    2. Update the function `hashOfConfigFor`, so that when the configuration object is hashed into a hash key and put into the eslintcache, if `shareableCache` is set to true, the file paths hashed are all relative paths. Implementation detail:
```

> ```

Reviewer 3: If you add `js` after the three ticks, we will get syntax highlighting.

function hashOfConfigFor(config, cwd) {
    if (!configHashCache.has(config)) {
        // replace the full directory path in the config string to make the hash location-neutral

>         const stringConfig = stringify(config).replaceAll(cwd, '.');

Reviewer 3: We should definitely be using `path.relative()` here rather than simple string replacement.

Author: I will be sure to update the implementation.

## Drawbacks

Adding a new CLI flag adds some complexity for the end user. They will need to understand why the feature exists. 

Furthermore, defaulting to the existing behavior does help make upgrading less painful, but it is possible most people expect a shareable cache by default. This may cause confusion.

## Backwards Compatibility Analysis

The new flag, shareable-cache, passed in on the command line, will default to false. Unless people explicitly set it to true, then the old algorithms will
be used. Users of the eslint cache will still need to regenerate their cache no matter how they set this flag, because the current implementation forces a cache rebuild if there are
any configuration changes. Adding a new property counts as a change. 

## Alternatives

A CLI tool which translates the existing cache to the current folder structure. This could be any sort of script that will basically do a replace-all on the paths in 
the .eslintcache file and also recalculates the hash so that they are the new location rather than the old. A downside to this is that any tool based on string 
replacement would be fragile at best. 

## Open Questions

No.

## Help Needed

> Decision needed: Make every cache shareable, or keep the shreable-cache flag default to false?

Reviewer 3: Note: all the questions in this section should actually be under "Open Questions"

Author: Moved.

I originally planned the shareable-cache flag so that users of eslint could not be surprised by changed cache behavior, and would also not have to regenerate their cache 
for a patch version change. But I discovered by testing that there is no way to NOT force people to regenerate their eslint cache with this change. That's because by adding 

> the 'shareable-cache' flag, it adds a property to the ConfigArray object. Even if it's set to false (the old behavior), the object's structure changes with a new property 

Reviewer 3: How is this changing the `ConfigArray`? This setting is at the CLI level, not at the config level, so there shouldn't be any changes to the `ConfigArray`.

field. This in turn changes the ConfigArray object's stringify results, which are fed into a hash function to create a hash key. (See lint-result-cache.js function hashOfConfigFor.) 
In turn, that becomes a part of the eslintcache file. I think it's so that everything will be re-scanned with configuration changes. 

> For that reason, should we simply have a shareable cache full of relative paths be the only behavior? That means no command line flag, and strictly internal changes that

Reviewer 3: We don't know how people are currently using the cache file, so we can't make assumptions about what is and isn't safe to change. Even if we made a shareable cache the default (which would be a breaking change), we'd still need a way to enable the previous behavior in case there is a use case we aren't aware of.

Author: Got it!  I will move it to questions and type in your answer.

- RFC PR: <https://github.com/eslint/rfcs/pull/106>
- Authors: Christian Schulz (@cschulz)i and cs6cs6

# Relative cache location strategy

## Summary

To be able to share the created eslint cache between different systems f.e. local machine and CI, there is a need for a relative path stored in the cache.

## Motivation

The overall motivation for doing this is performance during the eslint process.

Most people are running eslint on local developer machine and additional on the CI workers.
In high changing repositories there are several pull requests at the same time changing different files requiring linting all the time.
This is a time consuming process depending on the enabled rules like type checking rules.

If the CI worker is able to cache its last run, it can just pick this cache up and use it.
Also developers can reuse this cache or commit it as part of a repository for the CI.

## Detailed Design

I have a branch with the following design already. It is not fully tested, so I'm not 100% sure this will be the final implementation but I think it is a good start. 

The LintResultCache takes a file path as parameter to find or store a file cache entry.

> One approach would be to truncate the given absolute path of the file by the path of the cache location.

Author: Yes. I have tried to make the text clearer:

The suggested approach is to truncate the given absolute path of the file to a relative path in the cache.

### Adding the command line parameter

> - conf/default-cli-options.js: add the property 'shearableCache' with a default of false. It should be put in the section with the other cache variables, below cacheStrategy. 

Author: That was my first idea! But my concern was that leaked implementation details into the api, when people are actually interested in the functionality (semantic meaning).  If you're sure you prefer cachePaths, I am happy to change it.

- docs/src/use/command-line-interface.md: Add an explanation of the 'shareable-cache' variable with the other cache variables: "By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache."
- eslint-helpers.js: add shareableCache variable (set to false) to the processOptions
- lib/options.js: Add an option 'shareable-cache' of type Boolean with a nice description for people to read: By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache.

> 

Reviewer 3: It would be helpful to see the example file structure of the cache we use now vs. what it would look like if this RFC was implemented.

Author: This is a very good idea. I have added two sections showing the eslintcache file for a very tiny project. I did one section that is "now" and one section that is "Proposed"

### Changing cache file serialization
- lib/eslint/flat-eslint.js: Define the shearableCache property at the top, with the other properties, as a boolean
- lib/eslint/flat-eslint.js: Around line 873, where the flat-eslint.js file is saving the cache in the cachefile, if processOptions.shareableCache is set to true, pass a PARTIAL path to the lintResultCache.setCachedLintResults() function instead of the full path. 

## Documentation

Add the new CLI flag to the CLI documentation. The new cli flag will be called shareable-cache.

## Drawbacks

> No.

Author: Updated text:

Adding a new CLI flag adds some complexity for the end user. They will need to understand why the feature exists.

Furthermore, defaulting to the existing behavior does help make upgrading less painful, but it is possible most people expect a shareable cache by default. This may cause confusion.

## Backwards Compatibility Analysis

The new flag, shareable-cache, passed in on the command line, will default to false. Unless people explicitly set it to true, then the old algorithms will
be used. They will not need to regenerate their cache unless they set this flag. 

## Alternatives

> A CLI tool which translates the existing cache to the current folder structure.

Reviewer 3: Can you explain more about this? Would this be a separate tool from ESLint? What exactly would it do?

Author: A CLI tool which translates the existing cache to the current folder structure. This could be any sort of script that will basically do a replace-all on the paths in the .eslintcache file so that they are the new location rather than the old. A downside to this is that any tool based on string replacement would be fragile at best.

## Open Questions

No.

## Help Needed

> I have a branch with code changes, but I am trying to figure out how to build an artifact to test with. Using the eslint.js artifact in the browser does not make sense

Reviewer 3: You don't need to build anything. If you are just working locally, you can `npm install` the directory that you have ESLint in:

```
npm install ~/path/to/eslint
```

See https://docs.npmjs.com/cli/v10/commands/npm-install

Author: Thank you!  I got this working, and found things I missed in my first go-round. I'll be revising my design accordingly.
